sys/entropy_source/adc_noise/Kconfig

```diff
@@ -0,0 +1,74 @@
+# Copyright (c) 2020 HAW Hamburg
+#
+# This file is subject to the terms and conditions of the GNU Lesser
+# General Public License v2.1. See the file LICENSE in the top level
+# directory for more details.
+
+menuconfig KCONFIG_MODULE_ENTROPY_SOURCE_ADC_NOISE
+    bool "Configure ADC Noise entropy source module"
+    depends on MODULE_ENTROPY_SOURCE_ADC_NOISE
+    help
+        Configure the ADC Noise entropy source using Kconfig.
+
+if KCONFIG_MODULE_ENTROPY_SOURCE_ADC_NOISE
+
+choice
+    bool "ADC default sampling resolution"
+    default ENTROPY_SOURCE_ADC_RES_10BIT
+    help
+        This parameter sets the ADC sampling resolution. Please note that not all
+        platforms support every value.
+
+config ENTROPY_SOURCE_ADC_RES_6BIT
+    bool "6 Bit"
+
+config ENTROPY_SOURCE_ADC_RES_8BIT
+    bool "8 Bit"
+
+config ENTROPY_SOURCE_ADC_RES_10BIT
+    bool "10 Bit"
+
+config ENTROPY_SOURCE_ADC_RES_12BIT
+    bool "12 Bit"
+
+config ENTROPY_SOURCE_ADC_RES_14BIT
+    bool "14 Bit"
+
+config ENTROPY_SOURCE_ADC_RES_16BIT
+    bool "16 Bit"
+
+endchoice
+
+config ENTROPY_SOURCE_ADC_LINE_NUM
+    int "ADC line"
+    range 0 16
+    default 0
+    help
+        The ADC line maps to an I/O pin. This number acts as index to an array
+        of predefined ADC devices that contain the pin definition. Typically,
+        the array is defined by a board in a periph_conf.h file. Please note that
+        a board is not required to specify a minimum number of lines.
+
+config ENTROPY_SOURCE_ADC_HMIN
+    int "Min. entropy per sample [2^16 * bit/sample]"
+    range 1 524288
+    default 65536
+    help
+        Min. entropy needs to be evaluated thoroughly beforehand! To avoid float, the
+        actual entropy value per one byte sample needs to be manually multiplied by 2^16
+        before setting it as configuration parameter.
```

Comment:
According to the doc, this is a knob allowing to configure "the minimum sample per entropy". So this guarantees a lower bound of entropy that users of the API can expect this source of entropy to provide. This is a pretty wild claim, without any backing. And the simple example of a defect in the ADC proves that this claim is not universally true.

My "plethora of elaborations" was an attempt to expose this conceptional flaw.

Reply:
Then I don't understand you correctly. How would you solve this flaw? If I got it correctly, we agreed on a conservative ahead entropy estimation instead of a runtime estimation. This estimation to be (i) provided by the developer and (ii) configured during compile time. This is what the parameter is meant for.

Reply:
What is given here is the following guarantee: "This entropy source will under no circumstances provide less entropy, than this value". The minimum is after all the lowest value, thus an lower bound.

What I suggested to provide as ahead of time estimation is the following: "The chance of a random sample from a random entropy source of this type to contain at least the given amount of entropy is very high."

With the second claim, users have to expect that an entropy source might provide less entropy. In fact, a user would have to expect that (with a very small chance) an entropy sources might consistently provide little to no entropy at all.

Of course it would be nice to have a guaranteed lower bound of entropy instead. Maybe some special hardware entropy sources have been designed carefully enough and have been rigorously enough tested during production that an lower bound indeed can be given. But this is just not true for entropy sources in general; and especially not for ADCs. Thus, this is not a solid foundation to build the entropy system on top.

The second claim is (in theory) still good enough to be useful, as with mixing in more entropy sources and using more entropy samples the chance of the accumulated entropy actually contains the target quality increases. In fact, for every *p < 1* an *n* exists so that the probability of an entropy mix out of *n* entropy sources (while using at least so many samples of every entropy sources that according to the estimation each individually would be sufficient to fulfill the target entropy) contains at least the target entropy is at least *p*.

Reply:
> This entropy source will under no circumstances provide less entropy

Is it clear that this parameter needs to be **set** by the user aka firmware developer? Where to get the value from? From applying the NIST test suite. Something we could do in addition is multiplying this input value with a factor < 1 to be even more conservative about this estimation, like the Yarrow-160 CSPRNG does internally.

>The chance of a random sample from a random entropy source of this type to contain at least the given amount of entropy is very high

So would a documentation fix do?

Reply:
> So would a documentation fix do?

Yes.

But note that this has impact on the entropy accumulator, as it no longer can rely on a minimum entropy per sample. (Which I argue it couldn't in the first place, at least with impromptu entropy sources like the ADC, which do not undergo production tests to ensure that every specific piece of hardware performs well enough as entropy source. Even if the ADC would be rigorously tested, that check would likely not include if the ADC is usable for entropy harvesting.)

Reply:
@maribu can you take look at my latest commit? Let me known if you have further doc changes in mind.

Reply:
looks good to me